### PR TITLE
Mamba `slow_forward` gradient fix

### LINKED ISSUE #29514: Cannot propagate gradients in Mamba
### System Info

- `transformers` version: 4.39.0.dev0
- Platform: macOS-14.2.1-arm64-arm-64bit
- Python version: 3.11.7
- Huggingface_hub version: 0.21.4
- Safetensors version: 0.4.2
- Accelerate version: not installed
- Accelerate config: not found
- PyTorch version (GPU?): 2.2.0 (False)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: no
- Using distributed or parallel set-up in script?: no

### Who can help?

@ArthurZucker @younesbelkada 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

```python
from transformers import MambaConfig, MambaForCausalLM, AutoTokenizer
import torch

tokenizer = AutoTokenizer.from_pretrained("state-spaces/mamba-130m-hf")
model = MambaForCausalLM.from_pretrained("state-spaces/mamba-130m-hf")
inputs = tokenizer("Hey how are you doing?", return_tensors="pt")

def loss_fn(logits, target):
    return torch.nn.functional.cross_entropy(logits[:, -1, :], target)

outputs = model(**inputs)
loss = loss_fn(outputs.logits, torch.randn([1, 50280]))
loss.backward()
```

This produces the following error:

```bash
{
	"name": "RuntimeError",
	"message": "one of the variables needed for gradient computation has been modified by an inplace operation: [torch.FloatTensor [1, 1536, 16]], which is output 0 of torch::autograd::CopyBackwards, is at version 1; expected version 0 instead. Hint: enable anomaly detection to find the operation that failed to compute its gradient, with torch.autograd.set_detect_anomaly(True).",
	"stack": "---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Cell In[1], line 13
     11 outputs = model(**inputs)
     12 loss = loss_fn(outputs.logits, torch.randn([1, 50280]))
---> 13 loss.backward()

File ~/Documents/projects/inseq/.venv/lib/python3.11/site-packages/torch/_tensor.py:522, in Tensor.backward(self, gradient, retain_graph, create_graph, inputs)
    512 if has_torch_function_unary(self):
    513     return handle_torch_function(
    514         Tensor.backward,
    515         (self,),
   (...)
    520         inputs=inputs,
    521     )
--> 522 torch.autograd.backward(
    523     self, gradient, retain_graph, create_graph, inputs=inputs
    524 )

File ~/Documents/projects/inseq/.venv/lib/python3.11/site-packages/torch/autograd/__init__.py:266, in backward(tensors, grad_tensors, retain_graph, create_graph, grad_variables, inputs)
    261     retain_graph = create_graph
    263 # The reason we repeat the same comment below is that
    264 # some Python versions print out the first line of a multi-line function
    265 # calls in the traceback and some print out the last line
--> 266 Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
    267     tensors,
    268     grad_tensors_,
    269     retain_graph,
    270     create_graph,
    271     inputs,
    272     allow_unreachable=True,
    273     accumulate_grad=True,
    274 )

RuntimeError: one of the variables needed for gradient computation has been modified by an inplace operation: [torch.FloatTensor [1, 1536, 16]], which is output 0 of torch::autograd::CopyBackwards, is at version 1; expected version 0 instead. Hint: enable anomaly detection to find the operation that failed to compute its gradient, with torch.autograd.set_detect_anomaly(True)."
}
```

### Expected behavior

I would expect gradient backpropagation to work normally in the Mamba model, i.e. not to obtain an error upon calling `loss.backward()`

### PR DESCRIPTION
**Pull request by vasqu**
# What does this PR do?
- Fixes the gradient issue with mamba in the `slow_forward` by cloning the initial `ssm_state` obtained via the cache.
- Fixes typo in mamba test.
- Adds unused `create_and_check_forward_and_backwards` to the tests.
- Adds a `slow_forward` test (`test_mamba_cached_slow_forward_and_backwards`) checking if the cache doesn't break anything in the backward pass.

Fixes #29514 and duplicate #29526

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. See #29526
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@amyeroberts & @ArthurZucker

### CODE REVIEW

File: tests/models/mamba/test_modeling_mamba.py [hunk 1]
```diff
@@ -197,7 +197,25 @@ def create_and_check_state_equivalency(self, config, input_ids, *args):
...
+        self, config, input_ids, *args, gradient_checkpointing=False
+    ):
+        model = MambaModel(config)
+        model.to(torch_device)
+        if gradient_checkpointing:
+            model.gradient_checkpointing_enable()
+
+        # creating and using cache
+        _ = model(input_ids[:, :-1], use_cache=True)
+        outputs = model._slow_forward(input_ids[:, :-1], use_cache=True).last_hidden_state
```
Comment (amyeroberts) [inline tests/models/mamba/test_modeling_mamba.py:210]: I believe the method for the model isn't private. 

```suggestion
        outputs = model.slow_forward(input_ids[:, :-1], use_cache=True).last_hidden_state
```

t's funny this wasn't picked up in the tests. Can you confirm that you can run these tests and everything passes (i.e. they're not skipped)

**Review by amyeroberts — COMMENTED** [review]
Thanks for working on this! 

Change looks OK to me - although I have a question on whether the previous model relied on `cached_params.ssm_states` being updated in place. Could you run with and without this change to confirm the equivalent values on a forward pass? 

Other than that - just a question about the tests

**Comment by vasqu** [comment]
Hi @javiermcebrian, you won't need this fix if you have the `mamba-ssm` and `causal-conv1d` packages installed (including a GPU ofc). In case the packages are not recognized, you will get a warning and then this fix might be of interest. 

I can only speak for training and inference/generation, but for a private project (the "fast" variant), it worked fine.

File: tests/models/mamba/test_modeling_mamba.py (hunk 1 above)
Comment (vasqu) [reply tests/models/mamba/test_modeling_mamba.py]: I noticed that I used the torch module's `_slow_forward`, hence why it's "private". Guess I was doing this too hastily but the new commit should really use the mamba mixer `slow_forward` now. Hope, that makes it clearer. 

And here's a google colab [notebook](https://colab.research.google.com/drive/1kH6bqzylGSupJWBTA4qBxPCSl19c0lGJ?usp=sharing) showing that the generated tensors are indeed identical.

File: tests/models/mamba/test_modeling_mamba.py [hunk 2]
```diff
@@ -197,7 +197,29 @@ def create_and_check_state_equivalency(self, config, input_ids, *args):
...
+        if gradient_checkpointing:
+            model.gradient_checkpointing_enable()
+
+        # create cache
+        cache = model(input_ids, use_cache=True).cache_params
+        cache.seqlen_offset = 0
+
+        # use cache
+        token_emb = model.embeddings(input_ids)
+        outputs = model.layers[0].mixer.slow_forward(token_emb, cache)
```
Comment (amyeroberts) [inline tests/models/mamba/test_modeling_mamba.py:214]: We should also have a check on the shape of `outputs` here

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing this! 

Just a tiny comment to augment the test - otherwise looks good to me!

File: tests/models/mamba/test_modeling_mamba.py (hunk 2 above)
Comment (vasqu) [reply tests/models/mamba/test_modeling_mamba.py:214]: Hi @amyeroberts, I've added a shape assertion for the outputs to ensure the shape is preserved (through the mixer block, i.e. token_emb.shape == outputs.shape) but I'm not sure if that's what you had in mind. If not, just let me know.

Comment (amyeroberts) [reply tests/models/mamba/test_modeling_mamba.py:214]: I was thinking something like on L231, where we explicitly check for shape values against those set in the tester

Comment (vasqu) [reply tests/models/mamba/test_modeling_mamba.py:214]: Ah ok, I see. It should be changed to that now.

Comment (amyeroberts) [reply tests/models/mamba/test_modeling_mamba.py:214]: Thanks!

**Comment by amyeroberts** [comment]
All looks good to me - thanks for adding this! 

Let's just get a final review from @ArthurZucker to confirm behaviour is as desired for mamba and there isn't anything else that needs to be added

**Comment by ArthurZucker** [comment]
Reviewing now!

**Review by ArthurZucker — APPROVED** [review]
LGTM and sorry for the late review. Slow is slow anyway, that might slow it down even more but not a big deal.

**Pull request merged**